### PR TITLE
Create scaling objects instead of hpa

### PR DESCRIPTION
**Pull request by Richard87**

### CODE REVIEW

File: pkg/apis/deployment/scaledobject.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	if memoryTarget == nil && cpuTarget == nil {
+		cpuTarget = numbers.Int32Ptr(defaultTargetCPUUtilizationPercentage)
+	}
+
+	scaler := deploy.getScalerConfig(componentName, horizontalScaling.MinReplicas, horizontalScaling.MaxReplicas, cpuTarget, memoryTarget)
+
+	return deploy.kubeutil.ApplyScaledObject(ctx, namespace, scaler)
+}
+
+func (deploy *Deployment) garbageCollectDeprecatedHPAs(ctx context.Context) error {
```
Comment (nilsgstrabo) [inline pkg/apis/deployment/scaledobject.go:57]: This function will delete all HPAs, also HPAs controlled by ScaledObjects. We should only delete HPAs not controlled by ScaledObjects. We can do that by filtering on labels (set by ScaledOBject).
Example of HPA labels:
labels:
    app.kubernetes.io/managed-by: keda-operator
    app.kubernetes.io/name: keda-hpa-server
    app.kubernetes.io/part-of: server
    app.kubernetes.io/version: 2.14.0
    radix-app: radix-api
    radix-component: server
    scaledobject.keda.sh/name: server

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
Added some suggestions and comments.
We must also add ScaledObject to the radix-operator clusterrole (all verbs), and to the radix-api and radix-app-admin-envs clusterroles (get, list, watch) if radix-api needs to read the ScaledObject.

File: pkg/apis/utils/scaler.go [hunk 2]
```diff
@@ -14,3 +15,13 @@ func GetHpaMetric(hpa *v2.HorizontalPodAutoscaler, resourceName v1.ResourceName)
 	}
 	return nil
 }
+
+// GetScaleTrigger returns the trigger spec for the given name  and type or nil if not found
+func GetScaleTrigger(hpa *v1alpha1.ScaledObject, triggerName, triggerType string) *v1alpha1.ScaleTriggers {
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/utils/scaler.go:20]: Fixed

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 3]
```diff
@@ -1749,7 +1778,7 @@ func Test_HPA_Validation(t *testing.T) {
 				ra.Spec.Components[0].EnvironmentConfig[0].HorizontalScaling.MinReplicas = &minReplica
 			},
 			false,
-			false,
+			[]error{radixvalidators.ErrMinReplicasGreaterThanMaxReplicas, radixvalidators.ErrMinReplicasGreaterThanMaxReplicas},
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/radixvalidators/validate_ra_test.go]: Fixed, was a bug :)

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 4]
```diff
@@ -1589,24 +1589,24 @@ func Test_ValidationOfVolumeMounts_Errors(t *testing.T) {
...
+		updateRA updateRAFunc
+		isValid  bool
+		isErrors []error
 	}{
 		{
 			"horizontalScaling is not set",
 			func(ra *radixv1.RadixApplication) {},
 			true,
-			true,
+			[]error{},
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/radixvalidators/validate_ra_test.go:1601]: Nice! Fixed :)

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 5]
```diff
@@ -1589,24 +1589,24 @@ func Test_ValidationOfVolumeMounts_Errors(t *testing.T) {
...
 	var testScenarios = []struct {
-		name       string
-		updateRA   updateRAFunc
-		isValid    bool
-		isErrorNil bool
+		name     string
+		updateRA updateRAFunc
+		isValid  bool
+		isErrors []error
 	}{
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/radixvalidators/validate_ra_test.go:1596]: Agreed, with `isErrors = nil`, its pretty clear... fixed!

File: pkg/apis/deployment/kubedeployment.go [hunk 6]
```diff
@@ -21,15 +21,15 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-func (deploy *Deployment) createOrUpdateDeployment(ctx context.Context, deployComponent v1.RadixCommonDeployComponent) error {
+func (deploy *Deployment) recconcileDeployment(ctx context.Context, deployComponent v1.RadixCommonDeployComponent) error {
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/deployment/kubedeployment.go:24]: fixed

File: pkg/apis/deployment/scaledobject.go (hunk 1 above)
Comment (Richard87) [reply pkg/apis/deployment/scaledobject.go:57]: Fixed: Ignorerer HPA'en hvis `owner` ikke er `RadixDeployment`, da får vi slettet alle gamle hpas, men ikke de nye:

```golang
// If owner reference is *not* RadixDeployment, skip deleting it
if len(hpa.OwnerReferences) == 0 {
	continue
}
if owner := hpa.OwnerReferences[0]; owner.Kind != radixv1.KindRadixDeployment {
	continue
}
```

**Comment by Richard87** [comment]
> Added some suggestions and comments. We must also add ScaledObject to the radix-operator clusterrole (all verbs), and to the radix-api and radix-app-admin-envs clusterroles (get, list, watch) if radix-api needs to read the ScaledObject.

Fixed, added `get,list,watch` to app admins and app readers (in `-envs` cluster roles), and `get,list,create,update,patch,delete,watch` to  Radix Operator

File: pkg/apis/radixvalidators/validate_ra_test.go [hunk 7]
```diff
@@ -1731,15 +1759,16 @@ func Test_HPA_Validation(t *testing.T) {
...
 		},
 		{
 			"environment HPA minReplicas and maxReplicas are not set",
 			func(ra *radixv1.RadixApplication) {
 				ra.Spec.Components[0].EnvironmentConfig[0].HorizontalScaling = &radixv1.RadixHorizontalScaling{}
 			},
 			false,
-			false,
+			// TODO: Maybe this validation should not return ErrMinReplicasGreaterThanMaxReplicas (it happens because minReplicas defaults to 1
+			[]error{radixvalidators.ErrMinReplicasGreaterThanMaxReplicas, radixvalidators.ErrMinReplicasGreaterThanMaxReplicas},
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/radixvalidators/validate_ra_test.go]: Fixed, was a bug :)

File: pkg/apis/deployment/scaledobject.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+				kube.RadixAppLabel:       appName,
+				kube.RadixComponentLabel: componentName,
+			},
+			OwnerReferences: ownerReference,
+		},
+		Spec: kedav1.ScaledObjectSpec{
+			MinReplicaCount: minReplicas,
+			MaxReplicaCount: pointers.Ptr(maxReplicas),
+			PollingInterval: pointers.Ptr[int32](30), // Default
+			Advanced:        &kedav1.AdvancedConfig{RestoreToOriginalReplicaCount: true},
```
[the remark this answers is not in the document]
Comment (Richard87) [reply pkg/apis/deployment/scaledobject.go:133]: Will merge and release this before the Operator https://github.com/equinor/radix-api/pull/625

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**